src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  let prevState = { ...state };
+
+  for (const action of actions) {
+    let newState = { ...prevState };
+
+    switch (action.type) {
+      case 'addProperties':
+        for (const key in action.extraData) {
+          newState[key] = action.extraData[key];
+        }
+        break;
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+      case 'clear':
+        newState = {};
+        break;
+      default:
+        return Error('Bad action type');
```

Comment:
Instead of returning an Error object when the action type doesn't match any cases, consider throwing an error with 'throw new Error('Bad action type')'. Returning an Error object may lead to unexpected behavior as the function is expected to return an array of states.